GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -0,0 +1,162 @@
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public override void Dispose()
+        {
+            this.pipeClient?.Dispose();
+            base.Dispose();
+        }
+
+        internal static string CreateJsonMessage(
+            string vfsVersion,
+            string providerName,
+            string enlistmentId,
+            string mountId,
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            var message = new Dictionary<string, object>
+            {
+                ["version"] = vfsVersion,
+                ["providerName"] = providerName,
+                ["eventName"] = eventName,
+                ["eventLevel"] = ((int)level).ToString(),
+                ["eventOpcode"] = ((int)opcode).ToString(),
+                ["payload"] = new Dictionary<string, string>
+                {
+                    ["enlistmentId"] = enlistmentId,
+                    ["mountId"] = mountId,
+                    ["json"] = payload,
+                },
+
+                // TODO: do we need these?
+                // ETW-only properties
+                ["etw.activityId"] = activityId.ToString("D"),
+                ["etw.parentActivityId"] = parentActivityId.ToString("D"),
+            };
+
+            return JsonConvert.SerializeObject(message);
+        }
+
+        protected override void RecordMessageInternal(
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            string message = CreateJsonMessage(
+                this.vfsVersion,
+                this.providerName,
+                this.enlistmentId,
+                this.mountId,
+                eventName,
+                activityId,
+                parentActivityId,
+                level,
+                keywords,
+                opcode,
+                payload);
+
+            this.SendMessage(message);
+        }
+
+        private static string GetConfigValue(string gitBinRoot, string configKey)
+        {
+            string value = string.Empty;
+            string error;
+
+            GitProcess.ConfigResult result = GitProcess.GetFromSystemConfig(gitBinRoot, configKey);
+            if (!result.TryParseAsString(out value, out error, defaultValue: string.Empty) || string.IsNullOrWhiteSpace(value))
+            {
+                result = GitProcess.GetFromGlobalConfig(gitBinRoot, configKey);
+                result.TryParseAsString(out value, out error, defaultValue: string.Empty);
+            }
+
+            return value.TrimEnd('\r', '\n');
+        }
+
+        private void SendMessage(string message)

```

Comment:
This looks good for now.  (And I know I'm still working on the Mac version of the daemon, so some
of this is premature.)  But maybe (now or later) make this class abstract where it is here in the common
code and put this named-pipe SendMessage() routine in an override in Windows platform code.
(At least I think that is the way they do things in this code base.)